### ISSUE TITLE
Add support for global and site based stylesheets

### ISSUE THREAD

**Issue by evolve2k** [peterkeen/sites#9]
Add support for a global stylesheet that would let me default my sites to use something like Bootstrap or whatever I want as a base for all my sites. 

In addition to this add support for site based stylesheets (may already be there and maybe I just haven't got my head fully around your layout implementation yet).

Discussion Point: I'm getting a sense that philosophically, for an individual site almost that everything you need should be accessible and editable within the wiki gui interface using convention driven names like you have been doing. Global settings across all sites would be fine for the dev to keep in their fork of your repo. Is that how you're seeing it too?

**Comment by peterkeen** [comment]
So far my convention has been to commit  `/assets/bootstrap.min.css` with bootstrap and `/assets/main.css` with site-specific styles to the repo and then reference them with the `static_path` helper. For example:

```
<link href="<%= static_path '/assets/bootstrap.min.css' %>" rel="stylesheet">
<link href="<%= static_path '/assets/main.css' %>" rel="stylesheet">
```

`static_path` just prepends the given path with the `SCRIPT_NAME` variable, which in the case of viewing it inside the editor is `/:sitename/view`. Otherwise it's blank.

Maybe instead of global stylesheets we could have a sort of "template" site. Instead of setting up just a blank repo we could add a minimal layout, stylesheet, and `Home` wiki page when creating the site.

Re your discussion point, that's what I've been going for. The one place this falls down is that Gollum really doesn't like creating or editing files that have dots in the name, which makes it hard to edit, for example, a stylesheet.

**Comment by evolve2k** [comment]
I've not been having much luck getting the stylesheets rendering on my end using the static_path code.
Firstly I edited layout.erb to include a reference to the stylesheets as suggested above.

#File layout.erb

``` erb
<!doctype html>
<html>
    <head>
        <link href="<%= static_path('/assets/stylesheets/global.css')%>" rel="stylesheet" />
    </head>
  <body>
    <%= yield %>
  </body>
</html>
```

I monkeypatched the code to force it to render the layout, but it then reported that static_path method not found.
Reverting back I then went to play with the hard coded html in viewer.rb#get_layout as this is what it is rendering for me my default currently and it seems to be ignoring layout.erb all together.

I then attempted to edit the css string here just to see if I could at least get the css rendering somehow, but this just generated an obscure error along the lines of that the system wasn't happy with the provided html" 
(syntax error, unexpected '<') # within the html defined in the string

``` ruby
def get_layout(wiki)
      layout = wiki.file('layout.erb') || wiki.page('layout')
      if layout
        return layout.raw_data
      else
        return "<html>
          <head>
            <link href="<%= static_path('/assets/stylesheets/master.css') %>" rel="stylesheet" />
          </head>
          <body>
            <%= yield %>
          </body>
        </html>"
      end
    end
```

Where am I going wrong? 

Better yet, please commit a change to the code to include a working css reference within layout which points to say /assets/stylesheets/global.css and does something minor like making h1's blue or whatever.

Just reaching out after much head banging

**Comment by peterkeen** [comment]
@evolve2k I pushed the version that's running on `sites.bugsplat.info` up and also pushed an example repo: https://raw.github.com/peterkeen/docverter.com/master/layout.md

You can see my usage of `static_path` in a few places there.

**Comment by peterkeen** [comment]
@evolve2k I just pushed a change that will let you use the editor to create and edit `cs`, `js`, and `erb` files. The only small constraint is that you can't have css and js with the same name in the same directory. So, for example, you can't edit `/assets/main.css` and `/assets/main.js` with the editor. You can do that if you clone down the repo and edit files by hand, though.

https://github.com/peterkeen/sites/commit/57e2ad5d4a42a09f0a09b8ffb5487ae208f084e0

**Comment by evolve2k** [comment]
Happy New Year Peter, hope it's been a good one so far!

Thanks for the sample project and for making the changes above. 
The ability to edit css files etc within the gollums project is going to come in very useful indeed.

After running over this a few times I'm still stuck with getting the project to recognise css files which I place in the globally run layout.erb file.

Specifically, when I use:

```
# views/layouts.erb
<link href="<%= static_path '/bootstrap.min.css' %>" rel="stylesheet">
```

I'm getting the following error reported via the browser.

```
"NoMethodError at / undefined method `static_path' for #<Sites::Manager:...>"
```

Looking through the code I think this is the relevant section:

```
# viewer.rb - line 59
helpers do
  def static_path(path)
    env['SCRIPT_NAME'] + path
  end
end
```

I'm a little lost as to whether the issue is with the SCRIPT_NAME not being set or otherwise some issue with the helper method being called, or even if it's something else.

It seems I'm using static_path like you are, so just need some further guidance of what I'm not seeing.

Thanks Richie

**Comment by peterkeen** [comment]
Hm. Based on the error message it looks like you're trying to create a global view for all of your sites. Is that right?

That's not going to work, at least as-is. Each individual site has a `layout.erb` file in the root of the site repo.

**Comment by evolve2k** [comment]
Yep that's right, ahhh that's why it hasn't been working!

Ok I'll play around with site specific layout.erb files and give them a try but yes I'm needing to be able to create a global default view which applies to all my sites.

**Comment by evolve2k** [comment]
Per the README instructions: 
    "To create a site layout just create a wiki page named layout and set it to the erb type."

I've tried a few different approaches and getting different results, not sure which is correct.
Also there is no visual way to 'set it to erb type' except appending the word 'erb' somewhere which what I assume you mean. 

Approach 1: 
- Create a new page 'layout.erb'
- This then creates a wiki page called 'layout dot erb'

Approach 2: 
- Create a new page called 'layout'
- This then creates a wiki page called 'layout'
- Rename the page to 'layout.erb'

Which is the correct approach?

**Comment by peterkeen** [comment]
Approach 2, except you don't have to rename it. All you have to do is pick
"erb" in the edit mode dropdown on the right hand side of the editor
toolbar and it'll get the correct name.

**Comment by evolve2k** [comment]
I deleted everything I've tried before and tried as follows, still not seeing styles be applied.
- Create a new page called 'layout'
- This then creates a wiki page called 'layout'
- Change the dropdown to 'erb'
- Save and now have a page called 'layout'
- Create a new page called 'main'
- This then creates a wiki page called 'main'
- Change the dropdown to 'css'
- Save and now have a page called 'main'

Paste in example code from README.
My code is exactly per you example except I also added except:
1. Added a h1 of "I should be blue" within the body code as a quick test.
2. Your example code didn't include a "/head" tag so I added that too.

If I look under files per the gollum interface I have the following:
- Home
- layout
- main

Going to localhost:9292/myproject/view shows me content but is not applying any style.
Viewing source I see this within head:
link rel="stylesheet" href="/500_flinders_street/view/assets/stylesheets/main.css"
clicking this brings me a sinatra error "Sinatra doesn’t know this ditty."

**Comment by evolve2k** [comment]
Any thoughts what's up? 
Specifically when I follow the link to the stylesheet that should be linking through right?

**Comment by evolve2k** [comment]
Ok for anyone playing at home I've worked out a way to get things hooked up.
1. It seems the naming extension on the layout file does not matter. I created a page called 'layout' which when checked by looking at File is just called 'layout' no '.md' or '.erb' or anything like that.
2. The exact name you type when creating the wiki page for the css file is the exact same name you need to use in the related link tag in the layout file (obvious in hindsight).
   a. Create a new page # eg. 'main'
   b. On the edit page, select '.css' from the picklist if you like but this seemed not to make a difference for me.
   d. You can check in 'Files' and to ensure that the file is called just 'main'
3. Link to css file in the layout file.
   a. To link to files you use the following syntax: 

```
    <link href="<%= static_path '/main' %>" rel="stylesheet">
```

note that as the page was not renamed to be 'main.css' and is still just a file called 'main' that the link tag needs to refer to the same page name without the conventional .css at the end.  This I suppose is the usefulness of the rel="stylesheet" which mean that html is not relying on the file extension to know what to do. If you would like to refer to a file called .css then you need to:
1. Create the page as 'main' # or whatever
2. Rename the file and change it to 'main.css'
3. Then you can refer to main.css in the link tag, but this make no technical improvement to what is actually happening.

As you can see it has the following components, 
- Use of 'static_path' method within the link.
- reference to the file is to a file stored directly in the root directory eg. '/main'
- the file is referred to using the exact name you last gave it and as can be reviewed by looking in 'Files'.

Basically the way the file is named when you create the page needs to match what is used when you refer to it in the css. Seems obvious in hindsight, I think I was expecting some magic here when there was none.

So if you call the file 

```
'main' then use: <link href="<%= static_path '/main' %>" rel="stylesheet">
'main.css' then use: <link href="<%= static_path '/main.css' %>" rel="stylesheet">
```

The above does not address including a global stylesheet which applies across all projects, Peter has said this is feature is not currently available, it just resolves being able to wire up stylesheets within a project and specifically by doing everything through creating and editing pages within the Gollum wiki interface.

**Comment by evolve2k** [comment]
Nb. Addressing my earlier quadry on how the link tag reads when you view source once everything is linked up properly. After using static_path the link tag looks as follows when you view the source of the rendered page.

```
<link rel="stylesheet" href="/mysitename/view/main" />
```
